homeassistant/components/bme280/sensor.py

```diff
@@ -1,135 +1,142 @@
+        spi_bus = sensor_conf[CONF_SPI_BUS]
+        _LOGGER.info("BME280 sensor initialize at %s.%s", spi_bus, spi_dev)
+        sensor = await hass.async_add_executor_job(
+            partial(
+                BME280_spi,
+                t_mode=sensor_conf[CONF_OVERSAMPLING_TEMP],
+                p_mode=sensor_conf[CONF_OVERSAMPLING_PRES],
+                h_mode=sensor_conf[CONF_OVERSAMPLING_HUM],
+                standby=sensor_conf[CONF_T_STANDBY],
+                filter=sensor_conf[CONF_FILTER_MODE],
+                spi_bus=sensor_conf[CONF_SPI_BUS],
+                spi_dev=sensor_conf[CONF_SPI_DEV],
+            )
+        )
+        if not sensor.sample_ok:
+            _LOGGER.error("BME280 sensor not detected at %s.%s", spi_bus, spi_dev)
+            return False
+        sensor_handler = await hass.async_add_executor_job(
+            BME280Handler, sensor, INTERFACE_SPI
+        )
+    else:
+        i2c_address = sensor_conf[CONF_I2C_ADDRESS]
+        bus = smbus.SMBus(sensor_conf[CONF_I2C_BUS])
+        sensor = await hass.async_add_executor_job(
+            partial(
+                BME280_i2c,
+                bus,
+                i2c_address,
+                osrs_t=sensor_conf[CONF_OVERSAMPLING_TEMP],
+                osrs_p=sensor_conf[CONF_OVERSAMPLING_PRES],
+                osrs_h=sensor_conf[CONF_OVERSAMPLING_HUM],
+                mode=sensor_conf[CONF_OPERATION_MODE],
+                t_sb=sensor_conf[CONF_T_STANDBY],
+                filter_mode=sensor_conf[CONF_FILTER_MODE],
+                delta_temp=sensor_conf[CONF_DELTA_TEMP],
+                logger=_LOGGER,
+            )
+        )
+        if not sensor.sample_ok:
+            _LOGGER.error("BME280 sensor not detected at %s", i2c_address)
+            return False
+        sensor_handler = await hass.async_add_executor_job(
+            BME280Handler, sensor, INTERFACE_I2C
         )
-    )
-    if not sensor.sample_ok:
-        _LOGGER.error("BME280 sensor not detected at %s", i2c_address)
-        return False
-
-    sensor_handler = await hass.async_add_executor_job(BME280Handler, sensor)
 
     dev = []
-    with suppress(KeyError):
-        for variable in config[CONF_MONITORED_CONDITIONS]:
+    try:
+        for variable in sensor_conf[CONF_MONITORED_CONDITIONS]:
             dev.append(
                 BME280Sensor(sensor_handler, variable, SENSOR_TYPES[variable][1], name)
             )
+    except KeyError:
+        pass
 
     async_add_entities(dev, True)
 
 
 class BME280Handler:
-    """BME280 sensor working in i2C bus."""
+    """BME280 sensor working in SPI or I2C bus."""
 
-    def __init__(self, sensor):
+    def __init__(self, sensor, interface):
         """Initialize the sensor handler."""
         self.sensor = sensor
-        self.update(True)
+        self.interface = interface
+        if self.interface == INTERFACE_SPI:
+            self.update = self.update_spi
+            self.update()
```

Comment:
Please don't do side effects in init methods. Update the sensor before passing it to the handler.